imagenet/fp16util.py

```diff
@@ -0,0 +1,29 @@
+import torch
+import torch.nn as nn
+
+
+class tofp16(nn.Module):
+    def __init__(self):
+        super(tofp16, self).__init__()
+
+    def forward(self, input):
+        return input.half()
+
+
+def copy_in_params(net, params):
+    net_params = list(net.parameters())
+    for i in range(len(params)):
+        net_params[i].data.copy_(params[i].data)
+
+
+def set_grad(params, params_with_grad):
+
+    for param, param_w_grad in zip(params, params_with_grad):
+        if param.grad is None:
+            param.grad = torch.nn.Parameter(param.data.new().resize_(*param.data.size()))
+        param.grad.data.copy_(param_w_grad.grad.data)
+
+
+def replace_with_BN16():
```

Comment:
1) fp16 bn in core is currently broken
2) even if it were not, for convergence you need fp32 parameters and buffers.